File: practice_handout/nearest_neighbor/nearest_neighbor.js
// If you have an array of people described by name (string)
// and position (int). Each person has three friends, who
// are the three people with the closest position numbers
// to that person. Implement an algorithm that will find the
// three friends of a given person.

// Answer: sort the people by position numbers, then you only
// have to check the three people before and after a person.

var min = function(x,y) {
  if (x < y) {return x}
  return y;
}

var max = function(x,y) {
  if (x > y) {return x}
  return y;
}

var Person = function(name, position) {
  this.name = name;
  this.position = position;
  this.neighbors;
}



var mergeSort = function(array) {
  if (array.length <= 1) {return array}
  var mid = Math.floor(array.length / 2);
  var firstHalf = mergeSort(array.slice(0, mid));
  var secondHalf = mergeSort(array.slice(mid));
  return merge(firstHalf, secondHalf);
}

var merge = function(a, b) {
  var result = [];
  while (a.length > 0 && b.length > 0) {
    if (a[0] < b[0]) {
      result.push(a.shift());
    }
    else {result.push(b.shift());}
  }
  while (a.length > 0) {
    result.push(a.shift());
  }
  while (b.length > 0) {
    result.push(b.shift());
  }
  return result;
}

Array.prototype.eq = function(otherArray) {
  if (this.length !== otherArray.length) {
    return false;
  }
  for (var i in this) {
    if (this[i] !== otherArray[i]) {
      return false;
    }
  }
  return true;
}

console.log(mergeSort([1,4,3,2,5,3,1,1,7]).eq([1,1,1,2,3,3,4,5,7]));


a = new Person('Chris', 3)
b = new Person('Ross', 4)
c = new Person('Rachel', 3)
d = new Person('Matt', 10)
e = new Person('Julie', 7)
f = new Person('Sam', 8)

people = [a,b,c,d,e,f];

File: practice_handout/nearest_neighbor/nearest_neighbor.rb
# If you have an array of people described by name (string)
# and position (int). Each person has three friends, who
# are the three people with the closest position numbers
# to that person. Implement an algorithm that will find the
# three friends of a given person.

# Answer: sort the people by position numbers, then you only
# have to check the three people before and after a person.

Person = Struct.new(:name, :location, :neighbors) do
  def self.build(args)
    self.new(args[:name], args[:location], [])
  end
end

Neighborhood = Struct.new(:people) do
  def populateNeighbors
    ordered_people = people.sort {|x, y| x.location <=> y.location }
    find_neighbors(ordered_people)
  end

  def find_neighbors(array)
    array.each_with_index do |person, i|
      possible_neighbors = array.slice(max(0, i-4), 7)
      possible_neighbors.delete(person)
      while person.neighbors.length < 3
        neighbor = nearest_neighbor(possible_neighbors, person.location)
        person.neighbors << neighbor
        possible_neighbors.delete(neighbor)
      end
    end
  end

  def nearest_neighbor(possibles, location)
    possibles.sort{|x,y| (location - y.location).abs <=> (location - x.location).abs}.pop
  end

  def max(a,b)
    a > b ? a : b
  end
end


a = Person.build(name: 'Chris', location: 3)
b = Person.build(name: 'Rachel', location: 5)
c = Person.build(name: 'Matt', location: 4)
d = Person.build(name: 'Ross', location: 1)
e = Person.build(name: 'Spencer', location: 8)
f = Person.build(name: 'Katy', location: 2)
g = Person.build(name: 'Sam', location: 6)

neighborhood = Neighborhood.new([a,b,c,d,e,f,g])
p a.neighbors.empty?
neighborhood.populateNeighbors
p a.neighbors == [c,f,d]
